src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs

```diff
@@ -24,12 +24,19 @@ internal sealed class DistributedApplicationResourceBuilder<T>(IDistributedAppli
 
         // If the behavior is AddReplace then there should never be more than one annotation present. The following call will result in an exception which
         // allows us to easily spot these bugs.
-        if (behavior == ResourceAnnotationMutationBehavior.Replace && Resource.Annotations.OfType<TAnnotation>().SingleOrDefault() is { } existingAnnotation)
+        if (behavior == ResourceAnnotationMutationBehavior.Replace&& Resource.Annotations.SingleOrDefault(a => IsSameOrSubclassOf(a, annotation)) is { } existingAnnotation)
```

Comment:
You can use IsAssignableFrom - https://learn.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom?view=net-8.0

Neither type will ever be an interface so it should have the behavior you want - current class or sub class.